src/main/cpp/src/datetime_utils.cuh

```diff
@@ -137,6 +138,103 @@ struct date_time_utils {
     return (hour >= 0 && hour < 24) && (minute >= 0 && minute < 60) &&
            (second >= 0 && second < 60) && (microseconds >= 0 && microseconds < 1'000'000);
   }
+
+  /**
+   * @brief Returns day of week in civil calendar for the specified epoch days.
+   * Day of week is from 0(Monday) to 6(Sunday).
+   *
+   * @param epoch_days the days from epoch time 1970-01-01, it's int64 type,
+   * for Spark, only has int32 days, so [Int32.MinValue, Int32.MaxValue] is expected.
+   * @return the weekday for the days from epoch
+   */
+  __device__ static int64_t weekday_from_days(int64_t epoch_days) noexcept
+  {
+    // (Int32.MinValue - 8) days is Monday(value 0)
+    int64_t const min_monday_days =
+      static_cast<int64_t>(cuda::std::numeric_limits<int32_t>::min()) - 8L;
+    return (epoch_days - min_monday_days) % 7L;
```

Comment:
Does this work for really old dates too? Even with the switch in the calendars for very long ago dates? But I guess there wasn't DST then so there were no transitions. So can we at least add a note that this does not need to worry about really old dates because there are no transitions then.